### PR TITLE
Enemy Ball Placement Play + New Motion Constraint Validation Function

### PR DESCRIPTION
**Pull request by PhilD71**
### Description

This PR adds the Enemy Ball Placement Play used at Robocup 2021, now with proper validation functions to check that robots never enter the motion constraint created by ball placement. With this, a new non terminating validation function is introduced, `robotsNotViolatingMotionConstraints`. 
### Testing Done
Simulated tests written

### Resolved Issues

#1999 
### Length Justification

Its all one play + necessary adjustments to provide support for the new simulated tests
### Review Checklist

All done!
**_It is the reviewers responsibility to also make sure every item here has been covered_**

- [x] **Function & Class comments**: All function definitions (usually in the `.h` file) should have a javadoc style comment at the start of them. For examples, see the functions defined in `thunderbots/software/geom`. Similarly, all classes should have an associated Javadoc comment explaining the purpose of the class.
- [x] **Remove all commented out code**
- [x] **Remove extra print statements**: for example, those just used for testing
- [x] **Resolve all TODO's**: All `TODO` (or similar) statements should either be completed or associated with a github issue
- [x] **Justify drops in code coverage**: If this PR results in a non-trivial drop in code coverage (a bot should post a coverage diagram as a comment), please justify why we can't test the code that's not covered.

### CODE REVIEW

File: src/software/ai/hl/stp/play/enemy_ball_placement_play.h [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+
+   private:
+    /**
+     * Ball Placement tactic shadowing the defense enemy.
+     *
+     * @param crease_defenders list of crease defenders
+     * @param move_tactics list of move tactics
+     * @param placement_point ball placement point
+     * @param world world
+     * @param yield yield
```
Comment (akhilveeraghanta) [inline src/software/ai/hl/stp/play/enemy_ball_placement_play.h]: slight nit: order of params doesn't match the function

File: src/software/simulated_tests/non_terminating_validation_functions/robots_violating_motion_constraint.h [hunk 2]
```diff
@@ -0,0 +1,17 @@
...
+#include "software/ai/navigator/obstacle/obstacle.h"
+#include "software/ai/navigator/obstacle/robot_navigation_obstacle_factory.h"
+#include "software/simulated_tests/validation/validation_function.h"
+#include "software/world/world.h"
+
+/**
+ * Checks if the friendly robots are in the friendly half of the field, else assertion
+ * fails.
+ * @param world_ptr the world pointer given by the simulator. Gets updated every tick
+ * @param yield yields control to the next routine (coroutines) with error message
```
Comment (akhilveeraghanta) [inline src/software/simulated_tests/non_terminating_validation_functions/robots_violating_motion_constraint.h:13]: nit: missing obstacle_factory and constraints params

File: src/software/ai/hl/stp/play/enemy_ball_placement_play_test.cpp [hunk 3]
```diff
@@ -0,0 +1,217 @@
...
+
+            robotsViolatingMotionConstraint(
+                world_ptr, yield, obstacle_factory,
+                MotionConstraint::AVOID_BALL_PLACEMENT_INTERFERENCE);
+        }};
+
+    runTest(field, ball_state, friendly_robots, enemy_robots,
+            terminating_validation_functions, non_terminating_validation_functions,
+            Duration::fromSeconds(10));
+}
```
Comment (akhilveeraghanta) [inline src/software/ai/hl/stp/play/enemy_ball_placement_play_test.cpp]: These test seem to be the same code with just the ball position changing, you could reduce this down to 1 parameterized test

Checkout `simulated_chip_tactic_test.cpp` for an example, you can parameterize on the ball position

**Review by akhilveeraghanta — CHANGES_REQUESTED** [review]
Nice work! Just got a couple nits

File: src/software/ai/hl/stp/play/enemy_ball_placement_play.cpp [hunk 4]
```diff
@@ -0,0 +1,138 @@
...
+        }
+
+        // if there are threats, send one robot to placement point, and one shadows
+        else
+        {
+            move_tactics[0]->updateControlParams(
+                placement_point + placement_to_net,
+                placement_to_net.orientation() + Angle::half(), 0);
+            shadow_enemy->updateControlParams(enemy_threats.at(1),
+                                              ROBOT_MAX_RADIUS_METERS * 3);
```
Comment (jonathanlew) [inline src/software/ai/hl/stp/play/enemy_ball_placement_play.cpp]: how do we know there will be 2 enemy threats?

File: src/software/ai/hl/stp/play/enemy_ball_placement_play_test.cpp (hunk 3 above)
Comment (jonathanlew) [reply src/software/ai/hl/stp/play/enemy_ball_placement_play_test.cpp]: agreed, I'll re-review once it's cleaned up

File: src/software/geom/polygon.cpp [hunk 5]
```diff
@@ -119,6 +119,33 @@ Polygon Polygon::expand(const Vector& expansion_vector) const
     return Polygon(expanded_points);
 }
 
+Polygon Polygon::expand(double expansion_amount) const
```
Comment (jonathanlew) [inline src/software/geom/polygon.cpp:122]: is there a test for this new function?

**Review by jonathanlew — COMMENTED** [review]
just a few comments

File: src/software/ai/hl/stp/play/enemy_ball_placement_play.cpp (hunk 4 above)
Comment (PhilD71) [reply src/software/ai/hl/stp/play/enemy_ball_placement_play.cpp]: Typo - should be 0, fixed in next commit

File: src/software/geom/polygon.cpp (hunk 5 above)
Comment (PhilD71) [reply src/software/geom/polygon.cpp:122]: the tests got messed up when I untangled the branch from robocup 2021 - I have the new tests in now

File: src/software/simulated_tests/simulated_play_test_fixture.h [hunk 6]
```diff
@@ -51,6 +51,14 @@ class SimulatedPlayTestFixture : public SimulatedTestFixture
      */
     void setGameState(const GameState& game_state_);
 
+    /**
+     * Gets the configs used in simulation
+     * Useful for constructing duplicates of Obstacle Factory
+     *
+     * @return the Ai Config
+     */
+    std::shared_ptr<AiConfig> getAiConfig();
```
Comment (jonathanlew) [inline src/software/simulated_tests/simulated_play_test_fixture.h]: we shouldn't expose non-const pointers so it doesn't get modified elsewhere, i.e. the function signature should be `const std::shared_ptr<const AiConfig> getAiConfig() const`

**Review by jonathanlew — CHANGES_REQUESTED** [review]
looks good, just one comment

File: src/software/simulated_tests/simulated_play_test_fixture.h [hunk 7]
```diff
@@ -51,6 +51,14 @@ class SimulatedPlayTestFixture : public SimulatedTestFixture
      */
     void setGameState(const GameState& game_state_);
 
+    /**
+     * Gets the configs used in simulation
+     * Useful for constructing duplicates of Obstacle Factory
+     *
+     * @return the Ai Config
+     */
+    std::shared_ptr<AiConfig> getAiConfig() const;
```
Comment (jonathanlew) [inline src/software/simulated_tests/simulated_play_test_fixture.h]: same comment as before, the function signature should be `const std::shared_ptr<const AiConfig> getAiConfig() const`

**Review by akhilveeraghanta — COMMENTED** [review]
got nothing much to add beyond the existing comments, lgtm!

File: src/software/simulated_tests/simulated_play_test_fixture.h (hunk 7 above)
Comment (PhilD71) [reply src/software/simulated_tests/simulated_play_test_fixture.h]: oops should be fixed now

**Review by jonathanlew — APPROVED** [review]
LGTM!

**Review by akhilveeraghanta — APPROVED** [review]

**Pull request merged**